build: use `-muse-unaligned-vector-move` for Windows builds

fanquake opened the pull request:
We currently work around a longstanding GCC issue with aligned vector instructions, by patching the behaviour we want into GCC (see discussion in #24736). Possibly in response to the GCC thread (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54412#c40), a new option was [introduced into the binutils assembler](https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=c8480b58e1968f209b6365af7422678f348222c2) with the 2.38 release:
```
x86: Add -muse-unaligned-vector-move to assembler

Unaligned load/store instructions on aligned memory or register are as
fast as aligned load/store instructions on modern Intel processors.  Add
a command-line option, -muse-unaligned-vector-move, to x86 assembler to
encode encode aligned vector load/store instructions as unaligned
vector load/store instructions.
```

theuni commented:
Concept ACK. The patch we've taken from Debian is described as:

> replacing the aligned instruction with the unaligned one the hacky patch below got created, just replacing vmovapd by vmovupd. Not considering any side effects and maybe other instructions with alignment requirements.


Which appears to be what the new assembler option does as well. Using the supported switch is much better than the hacky patch.

What's up with the single aligned case though? Is that maybe coming from somewhere where we explicitly specify alignment?

theuni commented:
Any reason not to add it to our configure directly rather than guix?

hebasto commented:
Tested Guix build on Windows:

```
C:\Users\hebasto\Desktop\pr28151\bitcoin-6c3f6a9d015c-win64\bitcoin-6c3f6a9d015c>bin\bitcoind.exe -signet
*** stack smashing detected ***:  terminated

```

fanquake commented:
> Any reason not to add it to our configure directly rather than guix?

I think that's actually all we can do here, adding this (to configure) as a best-effort for anyone using an unpatched compiler (where it'd be most beneficial when building `bitcoind.exe`), while also retaining our GCC patching in Guix. Anything else would potentially leave us with unwanted code from dependencies/the rest of the toolchain.

fanquake committed:
build: use -muse-unaligned-vector-move for Windows

We currently work around a longstanding GCC issue with aligned vector
instructions, in our release builds, by patching the behaviour we want
into GCC (see discussion in #24736).

A new option now exists in the binutils assembler,
`-muse-unaligned-vector-move`, which should also achieve the behaviour
we want (at least for our code). This was added in the 2.38 release,
see
https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=c8480b58e1968f209b6365af7422678f348222c2.
```bash
x86: Add -muse-unaligned-vector-move to assembler

Unaligned load/store instructions on aligned memory or register are as
fast as aligned load/store instructions on modern Intel processors.  Add
a command-line option, -muse-unaligned-vector-move, to x86 assembler to
encode encode aligned vector load/store instructions as unaligned
vector load/store instructions.
```

Even if we introduce this option into our build system, we'll have to
maintain our GCC patching, as we want all code that ends up in the
binary, to avoid these instructions. However, there may be some value in
adding the option, as it could be an improvement for someone building
(bitcoind.exe) with an unpatched compiler.

fanquake commented:
> Edit: Ok, right, this is avx-only. Ignore most of the above.

Yea. It seems hard to (generally) produce working Windows binaries outside of Guix, because I assume most Windows toolchains are broken with the same bug (unless they are all being patched similar to Debian/Ubuntu).

So the best we can do is retain our GCC patching, so the entire toolchain/libs/bins are built correctly, and then in configure, we could pass the assembler flags as a best-effort, to _try_ and produce working binaries when building outside of Guix (and I don't think addition of the flags to the Guix build (from configure) would be an issue, as it would only be swapping out instructions that shouldn't be appearing in any case).

theuni commented:
ACK 96f2cf8d2c3e0fba2a39dabd991dee69124cc79d.

I verified locally that trying to use an assembler that doesn't understand the new option results in a compile failure:
```bash
 $ g++ -I. -std=c++17 -c test.cpp -o test.o -Wa,-muse-unaligned-vector-move || echo failure
as: unrecognized option '-muse-unaligned-vector-move'
failure
```

I also considered suggesting moving this to avx flags, but I guess it doesn't hurt to use it everywhere. It may end up having an effect on `sha-ni` code as well, for example.

fanquake commented:
> TODO:
Note that there is usage of vmovapd in this branch and the current release binaries. Need to follow up.

I might have either imagined this, or been looking at binaries that I'd broken, as I cannot seem to find any instructions in a Guix built ecab855838fa4de4c6d8c11e69037477d6047790. In any case, opened #28413, because we could add a test for the non-existence, while we continue to patch this behaviour out.
